### ISSUE TITLE
Question: Documentation or Types for Desktop Agent implementation of App Directory

### ISSUE THREAD

**Issue by Roaders** [finos/FDC3#1382]
### Question Area
 - [x] App Directory
 - [ ] API
 - [ ] Context Data
 - [ ] Intents
 - [ ] Use Cases
 - [ ] Other

### Question

We are close to implementing the app directory support for our Desktop Agent. I have been reading the docs but there does not seem to be any clear indication about how a Desktop Agent should handle app directory records - I am thinking specifically about how to open an app.

The only docs that I have found is the [swagger](https://fdc3.finos.org/schemas/2.1/app-directory.html#tag/Application/paths/~1v2~1apps~1%7BappId%7D/get). This is of some use but I was hoping for at least some Typescript `d.ts` files.

Is it specified anywhere how a DA should open an app?  I have made some assumptions:

- the DA should look through the hostManifests until it finds one it recognises and then use that to open the app.
- the host manifests should be listed in order of priority (if the DA knows about the first one it will open that and ignore the rest)
- ideally a web manifest should always be included at the end of the list so that a DA can always open an app even if know nothing about the other manifests
- the `url` supplied in the `details` section is not used for opening the app but is just used for app identity

Are these assumptions correct?
Thanks

**Comment by kriswest** [comment]
Hi @Roaders, prior to FDC3 2.0 there was simply a `manifest` element in appD records that was used to start an application. However, `manifest` was vendor-specific and you could only provide one. In 2.0's appD we replaced `manifest`:

- a _required_  `details` element expected to provide basic launch details that are vendor agnostic. It can contain a `url` property OR `path` and `arguments` for launching native apps, or `alias` and `arguments`  for citrix apps or other app asset systems using aliases, or `url` again for ClickOnce apps. 
- a _required_ `type` element that determines how the `details` element should be interpreted.
- a `hostManifests` object, intended to contain vendor-specific launch properties.

How a DA launches apps, using the details in the appD is largely left as an implementation detail. 

Regarding your assumptions:

1. >- the DA should look through the hostManifests until it finds one it recognises and then use that to open the app.

    Valid. However, it should also consider the `details` and `type` elements - these can be overridden by the hostManifest:

    > The manifest MAY duplicate or override information provided in the details field.

2. >- the host manifests should be listed in order of priority (if the DA knows about the first one it will open that and ignore the rest)

    Not valid - there is no implied ordering in the object. The manifests each have a key which should be used to identify manifests that you support, all others should be ignored.  The keys are not currently standardized (but should be IMHO), although a few examples are given. I think this is an area in which we could improve by standardizign keys - an issue should be raised for this.

3. >- ideally a web manifest should always be included at the end of the list so that a DA can always open an app even if know nothing about the other manifests

     As commented above this is not a list but an object. The example shows a `"Web App Manifest"` key that you can use to pick this up, but as commented above the keys are not properly standardized, which we could change easily enough.

4. >- the url supplied in the details section is not used for opening the app but is just used for app identity

    These elements predate any thoughts on app identity in FDC3 for the web and were absolutely intended to be used for launching apps, which is supported by the comments:
    
    > `type`: The technology type that is used to launch and run the application. Each application type implies a particular set of launch details. The supported types include:
    > - web: Web applications launched via a URL
    > - native: Native applications pre-installed on a device and launch via a filesystem path
    > - citrix: Apps virtualized via Citrixonline
    > - Native: Native apps that have an online launcher, e.g. online ClickOnce app deployments.
    > - other: Used to represent apps that do not conform to or cannot be launched via the other types, and are likely to be defined solely by a hostManifest.
    > FDC3 Desktop Agents MUST support at least the web application type and MAY support any or all of the other types.
    
    > `details`: The type specific launch details of the application. These details are intended to be vendor-agnostic and MAY be duplicated or overridden by details provided in the hostManifests object for a specific host.
    
    Rather FDC3 for the web is reusing the existing properties. There was also a suggestion that we should add an `allowedDomains` property to details to support identity validation (and multiple deployments/domains), but we never got around to adding it. If you (or any other participant from a fintech with on-prem deployments) ends up thinking that is needed it should be easy enough to add in.

If you are not seeing any of the comments quoted above in the appD spec, make sure you are clicking on the response elements, which pop open and provide the detail:
![image](https://github.com/user-attachments/assets/f50d5ead-8450-4e0f-ba47-e0cc0092f4bd)

Theres an issue open to create a better docs page using a docusaurus plugin: https://github.com/finos/FDC3/issues/1001

Although we should probably get into the docusaurus update first (to avoid having to repeat that): https://github.com/finos/FDC3/issues/1371

We could also add some content to the main overview on using the appD standard, which it currently lacks: https://fdc3.finos.org/docs/2.0/app-directory/overview
The last round of changes to that address use cases after these were reconfirmed, following a rejected issue that sought to remove appD from the standard...

I'd be happy to help improve this content if an issue is raised (or this converted into an issue to improve the docs).

**Comment by kriswest** [comment]
P.S. You are also correct that there are no generated or maintained TS types for appD records. The closest that we have is the [`AppMetadata`](https://fdc3.finos.org/docs/2.0/api/ref/Metadata#appmetadata) type which extends [`AppIdentifier`](https://fdc3.finos.org/docs/2.0/api/ref/Types#appidentifier) and is a stripped down set of the metadata from the AppD record. Personally, I don't think the reaction of info from the appD record was necessary, but past FDC3 maintainers had strong opinions on that. Personally I think the biggest omission from app metadata is a URL for the appD record! This can be determined from a fully qualified appId (see https://fdc3.finos.org/docs/2.0/app-directory/overview#shrinking-the-uri and https://fdc3.finos.org/docs/2.0/app-directory/overview#fully-qualified-appid-namespace-syntax-host-resolution) so I have never understood the lack of a more explicit/easier to use link to the record.

As the security discussion group is looking to make greater use of the appD records for signing and encryption, it may be time to address this issue (in FDC3 2.3?) and add an explicit link to the appD record used for launch into either `AppIdentifier` or `AppMetadata`. We could import the whole record into `AppMetadata` but I don't think that will be appropriate for the security group's use cases as they need the protection of the DNS system, rather than just going on info returned by the DA (as the use cases include protecting apps from nefarious DA implementations).

@Yannick-Malins @paulgoldsmith

**Comment by Roaders** [comment]
Thanks Kris

It seems like a good idea to provide more guidance on how a desktop agent should handle the app directory record. Ideally we would not have different behavior between different vendor implementations with the same app directory records. I am fine waiting for this to happen until after #1371 has been handled.

Would there be an objection to a PR adding types generated from the app directory schemas as we do for the browser types?

**Comment by kriswest** [comment]
To a certain extent there are differences in handling between agents baked into the concept of the `hostManifests` records - but we can manage that by starting to standardize the keys used (which seems long overdue). 

I see no issue, and several benefits to, creating some client code for working with app directories! There are a couple of related issues worth mentioning:
- #559
  - Client code should be able to deal with auth, however, there is no current reference implementation for either the server of client side of an appD. Although I do know of an implementation of an appD and could talk to the owner about contributing it...
- #563
  - As above - need that reference implementation.

(related as client code should come with tests against a reference implementation)

@robmoffat also created an implementation for the FINOS appD, but has commented that it might not be an ideal ref implementation.

Next, it might be worth thinking about the OpenAPI spec as a client for an appD can be generated directly from that, along with associated types I believe. E.g. with https://github.com/hey-api/openapi-ts
Alternatively, we could partition the OpenAPI spec into schemas for the records and an OpenAPI schema that imports them, then generate TS types from the record schemas to include in the FDC3 module (I suggest code generation as it prevents us having to maintain both schemas and matching types). Types can at least provide inline docs to developers working with appD records. We could also simply write some additional documentation... although I do prefer a 'single source of truth' so we avoid repeating ourselves/double maintenance.

We can ignore #1371, it shouldn't affect doing this.

Finally, it sounds like we need to start by adding some comments based on the above clarifications to the OpenAPI spec then agree what we are going to do with that... I'm happy to add this to the next SWG agenda and come up with a plan? If its something you want to start in on go for it - otherwise I'm happy to help.

**Comment by Roaders** [comment]
Thanks for the as always detailed reply @kriswest .

Yes, the `hostManifests` records allow differnet vendor implementations to behave differently but some guidance on what to do if multiple or no host manifests are known would be useful for example.

For the types, yes I was planning on using some type generation magic to generate them automatically from the schemas. Can we use the same mechanism as we do for `BrowserTypes` or is the source of truth different in this case. You mention the OpenAPI schema - that's still a json schema isn't it? Can't we use the same tools to convert that to TS as we do for `BrowserTypes`?

**Comment by kriswest** [comment]
@Roaders there is a historical relationship between OpenAPI and JSON schema - they are partially compatible in OpenAPI 3.0 (which appD is still specified in), but I think fully compatible in 3.1. More details here: https://apisyouwonthate.com/blog/openapi-v3-1-and-json-schema/

If we separate the schemas for the records out, the OpenAPI schema can import them. At that point we can use the same quicktype tooling to generate types for the record schemas. Generators for OpenAPI are different in that they generate REST server skeletons and REST clients (in many languages) which (I presume) also come with types for the API messages (the records returned and other responses). Hence, we have a couple of possible routes.

Lists of OpenAPI generators:
- https://tools.openapis.org/categories/code-generators.html
- https://openapi-generator.tech/docs/generators/

I guess the quesiton is, are full generated clients more useful or types for the AppD records for rolling your own client and handling code?

**Comment by robmoffat** [comment]
---
name: AppD Schema Centralization
overview: "Keep the App Directory specification in fdc3-standard (OpenAPI wrapper + JSON Schema record definitions), generate TypeScript via quicktype in @finos/fdc3-schema, migrate consumers, and add documentation that closes [issue #1382](https://github.com/finos/FDC3/issues/1382)."
todos:
  - id: split-schema
    content: Extract components.schemas from appd.schema.json into fdc3-standard/.../schemas/ JSON Schema hub + entry points; slim appd.schema.json to OpenAPI wrapper with external $refs; fix website copy-appd to copy schemas/ subdirectory
    status: in_progress
  - id: codegen
    content: Add quicktype typegen-appd to fdc3-schema via s2tQuicktypeUtil.cjs reading fdc3-standard JSON Schema files; generate AppDirectoryTypes.ts and export friendly aliases from src/index.ts
    status: pending
  - id: tests
    content: Update AppDirectory.test.ts to validate against JSON Schema Application definition; add generated-type compile test in fdc3-schema; verify ReDoc/Swagger still resolve external refs
    status: pending
  - id: migrate-web-impl
    content: Update fdc3-web-impl to import from @finos/fdc3-schema; remove local directory-openapi script and gitignored generated file
    status: pending
  - id: migrate-consumers
    content: Update demo and fdc3-example-apps to use shared Application type
    status: pending
  - id: docs
    content: "Add implementing-app-launch.md with launch algorithm and #1382 answers; use localhost-conformance.json as hostManifests worked example; update overview/spec links, CHANGELOG, AGENTS.md"
    status: pending
isProject: false
---

# App Directory Types via fdc3-schema

## Goal

Address [finos/FDC3#1382](https://github.com/finos/FDC3/issues/1382) by making App Directory record types a first-class, published artifact in [`packages/fdc3-schema`](packages/fdc3-schema), generated from schemas in [`packages/fdc3-standard`](packages/fdc3-standard). This replaces the current ad-hoc local generation in [`toolbox/fdc3-for-web/fdc3-web-impl`](toolbox/fdc3-for-web/fdc3-web-impl).

## Can we split appd.schema.json to use quicktype?

**Yes.** Quicktype only accepts JSON Schema draft-07 (`-s schema`), not OpenAPI 3.0. The current [`appd.schema.json`](packages/fdc3-standard/src/app-directory/specification/appd.schema.json) is an OpenAPI document with inline `components.schemas` — quicktype cannot consume it directly.

The solution is the same pattern already used for DACP: **separate record definitions (JSON Schema) from the API wrapper (OpenAPI)**.

```mermaid
flowchart TB
  subgraph fdc3standard [fdc3-standard - source of truth]
    OpenAPI["appd.schema.json\nOpenAPI: paths, servers, examples"]
    Hub["schemas/app-directory.schema.json\nJSON Schema draft-07 definitions hub"]
    Entry["schemas/application.schema.json\ntop-level entry for quicktype"]
    OpenAPI -->|"external $ref"| Hub
    Entry -->|"$ref #/definitions/Application"| Hub
  end
  subgraph fdc3schema [fdc3-schema - codegen]
    Quicktype["s2tQuicktypeUtil.cjs + quicktype"]
    Generated["generated/app-directory/AppDirectoryTypes.ts"]
    Entry --> Quicktype --> Generated
    Hub --> Quicktype
  end
```

### Recommended file layout (all under fdc3-standard)

```
packages/fdc3-standard/src/app-directory/specification/
  appd.schema.json                         # OpenAPI shell (paths, servers, security, examples)
  schemas/
    app-directory.schema.json              # definitions hub (~20 types: Application, Intent, Icon, …)
    application.schema.json                # top-level --src for quicktype
    all-applications-response.schema.json  # optional additional --src exports
  examples/application/                    # unchanged
```

**`app-directory.schema.json`** — extract current `components.schemas` into a draft-07 hub (mirrors [`api.schema.json`](packages/fdc3-schema/schemas/api/api.schema.json)):

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://fdc3.finos.org/schemas/next/app-directory/app-directory.schema.json",
  "definitions": {
    "Application": { ... },
    "Intent": { ... },
    "WebAppDetails": { ... }
  }
}
```

Internal refs change from `#/components/schemas/Foo` → `app-directory.schema.json#/definitions/Foo`.

**`application.schema.json`** — quicktype entry point (mirrors [`openRequest.schema.json`](packages/fdc3-schema/schemas/api/openRequest.schema.json)):

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://fdc3.finos.org/schemas/next/app-directory/application.schema.json",
  "title": "Application",
  "allOf": [{ "$ref": "app-directory.schema.json#/definitions/Application" }]
}
```

**`appd.schema.json`** — slim OpenAPI wrapper; each component becomes an external ref:

```json
"components": {
  "schemas": {
    "Application": { "$ref": "schemas/app-directory.schema.json#/definitions/Application" },
    "AllApplicationsResponse": { "$ref": "schemas/app-directory.schema.json#/definitions/AllApplicationsResponse" }
  },
  "examples": { ... }
}
```

Paths/responses continue to `$ref` `#/components/schemas/Application` as today.

### One-time extraction

Write a small script (or manual first pass) to:
1. Copy `components.schemas` → `definitions` in `app-directory.schema.json`
2. Rewrite all `#/components/schemas/X` → `app-directory.schema.json#/definitions/X`
3. Replace inline schemas in `appd.schema.json` with external `$ref`s
4. Validate with SwaggerParser (existing test) + JSON Schema example validation

No need to split into 20 individual files like DACP message schemas — the **hub + entry point** pattern is sufficient for ~20 record types.

### quicktype limitations (same as fdc3-context)

Document in generated output README / AGENTS.md:

| Schema construct | quicktype behaviour |
|-----------------|---------------------|
| `LaunchDetails` `anyOf` (Web/Native/Citrix/…) | Merged into one type with optional fields — usable but not a discriminated union |
| `HostManifests` `oneOf` (URI string \| object) | Merged — `{ [key: string]: string \| … }` |
| `HostManifest` open object | `Record<string, unknown>` or similar — expected |
| `Application` `allOf` | Generally handled correctly |

These are acceptable — same tradeoffs already documented in [`packages/fdc3-context/README.md`](packages/fdc3-context/README.md).

### Alternative considered: openapi-typescript

Works on the unsplit OpenAPI file but adds a second codegen toolchain. Splitting to JSON Schema lets AppD use the existing `s2tQuicktypeUtil.cjs` + quicktype pipeline — **preferred**.

### Alternative considered: dereference at build time

Run `@apidevtools/swagger-parser` to bundle OpenAPI → JSON Schema, then quicktype. Avoids manual split but produces an opaque generated intermediate and makes schema diffs harder to review. **Not preferred.**

## Design principle

**Schema source of truth stays in `fdc3-standard`.** Record definitions are JSON Schema; the OpenAPI file describes the REST API surface. `fdc3-schema` generates TypeScript via quicktype — same toolchain as DACP and context.

```
fdc3-standard/.../specification/schemas/   ← JSON Schema record definitions
fdc3-standard/.../specification/appd.schema.json   ← OpenAPI API wrapper
         │
         ▼  (quicktype via s2tQuicktypeUtil.cjs)
fdc3-schema/generated/app-directory/AppDirectoryTypes.ts
         │
         ▼
@finos/fdc3-schema exports Application, Intent, WebAppDetails, …
```

## Current state

- Canonical schema: single OpenAPI file with inline `components.schemas`
- Types generated only in `fdc3-web-impl` via `openapi-typescript` → gitignored `generated/directory-schema.ts`
- Tests in `packages/fdc3-standard/test/AppDirectory.test.ts`
- Website [`copy-appd`](website/package.json) copies only top-level `*.schema.json` — **will break ReDoc/Swagger after the split** (see below)

### Website `copy-appd` — must change with the split

Current script in [`website/package.json`](website/package.json):

```json
"copy-appd": "cpy \"../packages/fdc3-standard/src/app-directory/specification/*.schema.json\" \"static/schemas/next/\""
```

This glob copies **only** `appd.schema.json` from the specification root. It does **not** recurse into subdirectories.

After the split, `appd.schema.json` will `$ref` external files such as:

```json
"Application": { "$ref": "schemas/app-directory.schema.json#/definitions/Application" }
```

Those refs resolve relative to the published location (`static/schemas/next/appd.schema.json`), so the JSON Schema hub **must** land at `static/schemas/next/schemas/app-directory.schema.json`. Without updating `copy-appd`, ReDoc ([`app-directory.html`](website/static/schemas/2.2/app-directory.html) → `appd.schema.json`) and the [Swagger Editor link](website/docs/app-directory/spec.md) will fail to resolve `$ref`s.

**Proposed replacement:**

```json
"copy-appd": "cpy \"../packages/fdc3-standard/src/app-directory/specification/*.schema.json\" \"static/schemas/next/\" && cpy \"../packages/fdc3-standard/src/app-directory/specification/schemas/**\" \"static/schemas/next/schemas/\""
```

Also verify:
- [`version:schemas-replace`](website/package.json) — replaces `/schemas/next` in `**/*.schema.json`; should cover new files under `static/schemas/next/schemas/`
- [`version:appd-replace`](website/package.json) — only touches `appd.schema.json`; still sufficient for OpenAPI version field
- [`website/src/pages/schemas/next/app-directory.html`](website/src/pages/schemas/next/app-directory.html) — currently points at stale `app-directory.yaml`; static 2.2 HTML correctly uses `appd.schema.json`. Fix the `next` page as part of this work (optional but recommended)

Compare with [`copy-schemas`](website/package.json), which already copies recursively:

```json
"copy-schemas": "... cpy \"../packages/fdc3-schema/schemas/**\" static/schemas/next/ ..."
```

## Phase 1 — Split schema in fdc3-standard

1. Create `specification/schemas/app-directory.schema.json` with all record definitions
2. Create `specification/schemas/application.schema.json` (and optionally `all-applications-response.schema.json`) as quicktype entry points
3. Slim `appd.schema.json` to external `$ref`s for all `components.schemas`
4. Keep `components.examples` in the OpenAPI file (ReDoc examples)
5. **Fix website publishing** — update `copy-appd` (see dedicated section above); verify ReDoc and Swagger Editor resolve external refs after `npm run prebuild`
6. Verify SwaggerParser validation still passes

## Phase 2 — Code generation in fdc3-schema

Add to [`packages/fdc3-schema/package.json`](packages/fdc3-schema/package.json):

```json
"typegen-appd": "node s2tQuicktypeUtil.cjs ../fdc3-standard/src/app-directory/specification/schemas/application.schema.json ../fdc3-standard/src/app-directory/specification/schemas/app-directory.schema.json app-directory ../generated/app-directory/AppDirectoryTypes.ts"
```

(Exact arguments follow the same convention as `typegen-browser` — pass entry `.schema.json` files as `--src` and the hub as `-S`.)

Integrate into `generate` → `build`. **Commit generated output.**

Add exports in [`packages/fdc3-schema/src/index.ts`](packages/fdc3-schema/src/index.ts):

```typescript
export type { Application, Intent, WebAppDetails, HostManifests, Interop, AllApplicationsResponse } from '../generated/app-directory/AppDirectoryTypes.js';
```

No new devDependency — reuses existing quicktype.

## Phase 3 — Tests

**Update in `fdc3-standard`:**
- SwaggerParser validates composed `appd.schema.json` (with external refs resolved)
- Example JSON validates against `Application` definition (load from `app-directory.schema.json#/definitions/Application`)

**Add in `fdc3-schema`:**
- `tsc` confirms generated AppD types compile

## Phase 4 — Migrate consumers

- **fdc3-web-impl**: remove `directory-openapi`, import from `@finos/fdc3-schema`
- **demo**: type `hostManifests.demo` using generated types
- **fdc3-example-apps**: replace inline `AppDirectoryRecord` with `Application`

Hand-written API types (`AppMetadata`, `Icon`, `Image`) stay in `fdc3-standard/src/api/`.

## Phase 5 — Documentation (closes #1382)

Add [`website/docs/app-directory/implementing-app-launch.md`](website/docs/app-directory/implementing-app-launch.md) with:

1. TypeScript imports from `@finos/fdc3-schema`
2. Launch algorithm (flowchart)
3. **Understanding `hostManifests`** — practical table (host keys, details vs hostManifests, override semantics, no ordering)
4. Answers to Roaders' #1382 assumptions
5. Worked example from [`localhost-conformance.json`](toolbox/fdc3-example-apps/directory/static/localhost-conformance.json) (`Conformance1`, `ChannelsAppId`, link to `dummy-desktop-agent.ts`)
6. `Application` vs `AppMetadata` mapping

Link from overview and spec. Update CHANGELOG and AGENTS.md.

## Phase 6 — PR strategy

1. **Schema split** in fdc3-standard + website copy fix + updated tests
2. **quicktype codegen + exports** in fdc3-schema
3. **Consumer migration** (fdc3-web-impl, demo, example-apps)
4. **Documentation** + issue #1382 closure

## Risks

| Risk | Mitigation |
|------|------------|
| ReDoc/Swagger break on external refs | Update `copy-appd` to copy `schemas/` subfolder; add SwaggerParser test with dereferencing; fix stale `next/app-directory.html` |
| quicktype merges `anyOf`/`oneOf` | Document known limitation (same as fdc3-context); types remain usable |
| `HostManifest` open object | Doc + `Record<string, unknown>`; patterns in localhost-conformance.json |
| Duplicate Icon/Image vs fdc3-standard API types | Document relationship; do not merge |
| Schema split is a large diff | One-time extraction script; single PR for split |

## Success criteria

- Record definitions live as JSON Schema under `fdc3-standard/.../specification/schemas/`
- Website `copy-appd` updated; ReDoc/Swagger resolve external `$ref`s on fdc3.finos.org
- Published `@finos/fdc3-schema` exports `Application` and related types via **quicktype**
- `fdc3-web-impl` uses shared types; local openapi-typescript generation removed
- Launch documentation answers #1382 with practical `hostManifests` guidance
- Example records validated in CI
- CHANGELOG + AGENTS.md updated

**Comment by robmoffat** [comment]
ready for review
